DM-45159: Add debug output for Butler queries

dhirving opened the pull request:
Add an environment variable DAF_BUTLER_DEBUG_QUERIES that enables logging of queries and query plans.

## Checklist

- [x] ran Jenkins
- [ ] added a release note for user-visible changes to `doc/changes`
- [ ] (if changing dimensions.yaml) make a copy of dimensions.yaml in `configs/old_dimensions`

dhirving committed:
Add debug output for Butler queries

Add an environment variable DAF_BUTLER_DEBUG_QUERIES that enables logging of queries and query plans.

andy-slac commented on python/lsst/daf/butler/registry/interfaces/_database_explain.py:
```diff
@@ -0,0 +1,79 @@
+
+@compiles(_Explain, "postgresql")
+def _compile_explain(element: _Explain, compiler: SQLCompiler, **kw: Any) -> str:
+    text = "EXPLAIN "
```
It may be useful to add `(VERBOSE TRUE, SETTINGS TRUE)` after `EXPLAIN` for extra information.

andy-slac commented on python/lsst/daf/butler/registry/interfaces/_database_explain.py:
```diff
@@ -0,0 +1,79 @@
+
+    Parameters
+    ----------
+    statement : `sqlalchemy.SelectBase`
```
Annotation says `ClauseElement`?

andy-slac commented on python/lsst/daf/butler/registry/interfaces/_database.py:
```diff
@@ -1875,6 +1879,25 @@ def query(
+        the environment variable DAF_BUTLER_DEBUG_QUERIES is set to a truthy
+        value.
+        """
+        if os.environ.get("DAF_BUTLER_DEBUG_QUERIES", False):
```
Can we do this once only?

dhirving replied:
It's a dict lookup, the network I/O and disk I/O surrounding it is going to dwarf the execution time by several orders of magnitude.

Also in notebooks it's nice to be able to twiddle the variable after startup.

andy-slac commented on python/lsst/daf/butler/registry/interfaces/_database.py:
```diff
@@ -1875,6 +1879,25 @@ def query(
+            )
+
+            query_plan = get_query_plan(connection, sql)
+            print(f"Query plan:\n{query_plan}", file=sys.stderr)
```
Maybe send this to `logging`?

dhirving replied:
Yeah I considered that but then to use it you'd have to also figure out how to get logging turned on, so it seemed easier to just send it to stderr.  Not opposed to it if you want to make that change yourself though.

andy-slac reviewed (approved):
Looks great, few minor questions. Would it be useful to add timing information for actual query execution to debugging output?

dhirving committed:
Make query explain more verbose

Andy suggested some useful options that give you more information about the query plan.
